File: backend/guardian/routers/trips.py
from fastapi import APIRouter, Header, HTTPException
from pydantic import BaseModel

from guardian.config import has_insforge
from guardian.services.insforge_client import InsForgeClient, InsForgeError

router = APIRouter()


class TripBody(BaseModel):
    name: str
    origin: str
    destination: str
    mode: str = "walking"


def _token_from_header(authorization: str | None) -> str | None:
    if not authorization:
        return None
    return authorization.removeprefix("Bearer ").strip() or None


@router.get("")
async def list_trips(authorization: str | None = Header(None)):
    if not has_insforge():
        raise HTTPException(503, "InsForge not configured")
    token = _token_from_header(authorization)
    try:
        client = InsForgeClient(token=token) if token else InsForgeClient()
        trips = await client.list_trips()
        return {"trips": trips}
    except InsForgeError as e:
        raise HTTPException(400, str(e)) from e


@router.post("")
async def create_trip(body: TripBody, authorization: str | None = Header(None)):
    if not has_insforge():
        raise HTTPException(503, "InsForge not configured")
    token = _token_from_header(authorization)
    try:
        client = InsForgeClient(token=token) if token else InsForgeClient()
        user = await client.get_current_user()
        user_id = (user or {}).get("id")
        trip = await client.insert_trip(
            {
                "name": body.name,
                "origin": body.origin,
                "destination": body.destination,
                "mode": body.mode,
                "user_id": user_id,
            }
        )
        return {"trip": trip}
    except InsForgeError as e:
        raise HTTPException(400, str(e)) from e

File: backend/src/routes/trips.ts
import { Router } from "express";
import { createClient } from "@insforge/sdk";
import { config, hasInsforge } from "../config.js";

export const tripsRouter = Router();

function clientFromReq(req: { headers: { authorization?: string } }) {
  const token = req.headers.authorization?.replace(/^Bearer\s+/i, "");
  return createClient({
    baseUrl: config.insforge.url,
    anonKey: config.insforge.anonKey,
    ...(token ? { accessToken: token } : {}),
  });
}

tripsRouter.get("/", async (req, res) => {
  if (!hasInsforge()) {
    res.status(503).json({ error: "InsForge not configured" });
    return;
  }
  const client = clientFromReq(req);
  const { data, error } = await client.database.from("trips").select("*").order("created_at", { ascending: false });
  if (error) {
    res.status(400).json({ error: error.message });
    return;
  }
  res.json({ trips: data ?? [] });
});

tripsRouter.post("/", async (req, res) => {
  if (!hasInsforge()) {
    res.status(503).json({ error: "InsForge not configured" });
    return;
  }
  const { name, origin, destination, mode = "walking" } = req.body ?? {};
  const client = clientFromReq(req);
  const { data: userData } = await client.auth.getCurrentUser();
  const userId = userData?.user?.id;
  const { data, error } = await client.database
    .from("trips")
    .insert([{ name, origin, destination, mode, user_id: userId }])
    .select()
    .single();
  if (error) {
    res.status(400).json({ error: error.message });
    return;
  }
  res.json({ trip: data });
});
